x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/MlAutoscalingResourceTracker.java

```diff
@@ -208,47 +219,118 @@ static void getMemoryAndProcessors(
             final int numberOfThreadsPerAllocation = assignment.getTaskParams().getThreadsPerAllocation();
             final long estimatedMemoryUsage = assignment.getTaskParams().estimateMemoryUsageBytes();
+            final int numTargetAllocationsOnExistingNodes = assignment.getNodeRoutingTable()
+                .values()
+                .stream()
+                .mapToInt(RoutingInfo::getTargetAllocations)
+                .sum();
+            final int numMissingAllocations = numberOfRequestedAllocations - numTargetAllocationsOnExistingNodes;
+            final int numProcessorsNeeded = numMissingAllocations * numberOfThreadsPerAllocation;
+            int numberOfAvailableProcessors = numberOfAvailableProcessors(
+                autoscalingContext,
+                allocatedProcessorsScale,
+                processorsNewlyAssignedToTrainedModelsAllocationsThisTime
+            );
+            int numExistingProcessorsToBeUsed = Math.min(numProcessorsNeeded, numberOfAvailableProcessors);
 
             if (AssignmentState.STARTING.equals(assignment.getAssignmentState()) && assignment.getNodeRoutingTable().isEmpty()) {
-
+                // this state occurs when the assignment has just been added and has not been assigned to any nodes
                 logger.debug(
                     () -> format(
                         "trained model [%s] lacks assignment , memory required [%d]",
                         modelAssignment.getKey(),
                         estimatedMemoryUsage
                     )
                 );
-
-                extraSingleNodeModelMemoryInBytes = Math.max(extraSingleNodeModelMemoryInBytes, estimatedMemoryUsage);
+                extraPerNodeModelMemoryBytes = Math.max(extraPerNodeModelMemoryBytes, estimatedMemoryUsage);
                 extraModelMemoryInBytes += estimatedMemoryUsage;
 
-                // if not low priority, check processor requirements
+                // if not low priority, check processor requirements. If
                 if (Priority.LOW.equals(modelAssignment.getValue().getTaskParams().getPriority()) == false) {
-                    // as assignments can be placed on different nodes, we only need numberOfThreadsPerAllocation here
-                    extraSingleNodeProcessors = Math.max(extraSingleNodeProcessors, numberOfThreadsPerAllocation);
-                    extraProcessors += numberOfAllocations * numberOfThreadsPerAllocation;
+                    if (numProcessorsNeeded > numberOfAvailableProcessors) {
+                        // as assignments can be placed on different nodes, we only need numberOfThreadsPerAllocation here
+                        extraSingleNodeProcessors = Math.max(extraSingleNodeProcessors, numberOfThreadsPerAllocation);
+                        extraProcessors += numberOfRequestedAllocations * numberOfThreadsPerAllocation - numExistingProcessorsToBeUsed;
```

Comment:
I think thats right. I think in this case the total number of allocations should be the same as the missing allocations because we are just STARTING, but it would be clearer to use the same variables everywhere.

the memory is added just above this line, I believe